docs/governance/rfc/005-python-implementation.md

```diff
@@ -0,0 +1,134 @@
+  protocols:
+    - PROTOCOL_CONNECT
+  codecs:
+    - CODEC_PROTO
+    - CODEC_JSON
+  compressions:
+    - COMPRESSION_IDENTITY
+    - COMPRESSION_GZIP
+    - COMPRESSION_BR
+    - COMPRESSION_ZSTD
+  stream_types:
+    - STREAM_TYPE_UNARY
+    - STREAM_TYPE_CLIENT_STREAM
+    - STREAM_TYPE_SERVER_STREAM
+    - STREAM_TYPE_HALF_DUPLEX_BIDI_STREAM
+  supports_half_duplex_bidi_over_http1: true
+  supports_h2c: false
+```
+
+## General design
+
+Most of the code for both servers and clients will be in a connect-python runtime library, installable with `pip install connect-python`.
+
+A code generator, written in Python and installable with `pip install connect-python[compiler]`, will be available to generate a few functions:
+
+1. Synchronous and asynchronous client constructors
+2. Synchronous and asynchronous `typing.Protocol` that defines the method set that server implementations must follow
+3. Synchronous server constructor that produces a WSGI application given an implementation
+4. Asynchronous server constructor that produces an ASGI application given an implementation
+
+### Full type hinting, checked by mypy
+
+We will make sure the runtime, and all generated code, pass mypy type checks in strict mode.
+
+### Opt-in compression
+
+Compression codecs that are not in the Python standard library (Brotli and zstd) should be supported if their associated libraries are present, but those libraries should not be direct dependencies.
+
+### High- and low-level client APIs
+
+Connect clients are able to send arbitrary headers, and inspect headers sent by servers.
+In certain circumstances, they can also inspect or send trailers.
+Full support of this is enforced by the conformance test suite.
+
+But these advanced features can cloud the API for typical usage.
+Headers and trailers are unused in the majority of RPC usage; the RPC's message types typically encode all the relevant information.
+
+So, we will provide clients with both high-level simple methods which operate directly on the request and response types, and also low-level methods that wrap request and response types in classes that permit handling of headers and trailers.
+
+The low-level RPCs will be named `call_<rpc>`. For example, `call_converse` or `call_say` for an `rpc Converse` or `rpc Say`.
+The high-level RPCs will be simply named `<rpc>`. For example, `converse` or `say`.
+
+To make this more concrete, here is a portion of the proposed method set of the synchronous Eliza client:
+```python
+class ElizaServiceClient:
+	def say(self, req: SayRequest) -> SayResponse:
+		...
+	
+	def call_say(self, req: SayRequest, extra_headers: MultiDict[str, str], timeout_seconds: float) -> UnaryOutput[SayResponse]:
+		...
+		
+	def converse(self, reqs: Iterable[ConverseRequest]) -> Iterator[ConverseResponse]:
+		...
+		
+	def call_converse(self, reqs: Iterable[ConverseRequest], extra_headers: MultiDict[str, str], timeout_seconds: float) -> StreamOutput[ConverseResponse]:
+		...
+```
```

Comment:
Maybe this can be deferred to the implementation, but can't we unify them with `**kwargs`? I haven't written any python so I apologize if I am missing something obvious.

Reply:
Oh, I should also note - the response types are different, so you'd still need separate methods.

Reply:
> the response types are different, so you'd still need separate methods.

They need not be. We can use callbacks for that. But this can deferred to the implementation, no need to resolve in the RFC.

Reply:
Yes, we'd raise exceptions to let you look at the error. I agree that error code handling is important.

I think you're right that optional positional parameters could be a good fit here. The response type the more difficult nut to crack. 

In practice, I've found the high-vs-low API distinction useful largely because I use the high-level API exclusively in all my real production code, and use the low-level API to pass conformance tests. I just never care about trailers myself, but conformance tests demand we expose them somehow.